### PR TITLE
TWILL-107 Add payload support for Discoverable

### PR DESCRIPTION
**Pull request by gokulavasan**
JIRA : https://issues.apache.org/jira/browse/TWILL-107

### CODE REVIEW

File: twill-discovery-api/src/main/java/org/apache/twill/discovery/Discoverable.java [hunk 1]
```diff
@@ -19,19 +19,67 @@
...
 /**
  * Discoverable defines the attributes of service to be discovered.
  */
-public interface Discoverable {
+public class Discoverable {
+  private final String name;
+  private final InetSocketAddress address;
+  private final byte[] payload;
 
+  public Discoverable(String name, InetSocketAddress address, byte[] payload) {
```
Comment (chtyim) [inline twill-discovery-api/src/main/java/org/apache/twill/discovery/Discoverable.java]: Should also have a constructor that doesn't take `payload`.

File: twill-discovery-api/src/main/java/org/apache/twill/discovery/Discoverable.java [hunk 2]
```diff
@@ -19,19 +19,67 @@
...
    */
-  InetSocketAddress getSocketAddress();
+  public InetSocketAddress getSocketAddress() {
+    return address;
+  }
+
+  /**
+   * @return A payload represented as a byte array
+   */
+  byte[] getPayload() {
```
Comment (chtyim) [inline twill-discovery-api/src/main/java/org/apache/twill/discovery/Discoverable.java]: Should this be a `public` method?

File: twill-discovery-api/src/main/java/org/apache/twill/discovery/Discoverable.java [hunk 3]
```diff
@@ -19,19 +19,67 @@
...
+
+    Discoverable other = (Discoverable) o;
+
+    return name.equals(other.getName()) && address.equals(other.getSocketAddress()) &&
+      Arrays.equals(payload, other.getPayload());
+  }
+
+  @Override
+  public int hashCode() {
+    int result = name.hashCode();
```
Comment (chtyim) [inline twill-discovery-api/src/main/java/org/apache/twill/discovery/Discoverable.java]: Use `Objects.hash(name, address, payload)` instead.

File: twill-discovery-core/src/main/java/org/apache/twill/discovery/DiscoverableAdapter.java [hunk 4]
```diff
@@ -76,18 +78,9 @@ public Discoverable deserialize(JsonElement json, Type typeOfT,
       final String service = jsonObj.get("service").getAsString();
       String hostname = jsonObj.get("hostname").getAsString();
       int port = jsonObj.get("port").getAsInt();
+      final byte[] payload = context.deserialize(jsonObj.get("payload"), BYTE_ARRAY_TYPE);
```
Comment (chtyim) [inline twill-discovery-core/src/main/java/org/apache/twill/discovery/DiscoverableAdapter.java]: No need to be final. Same for `service` above.

**Comment by chtyim** [comment]
Just couple comments.

**Comment by gokulavasan** [comment]
Thanks for the review @chtyim. Addressed comments. Please take a look again.

**Comment by chtyim** [comment]
LGTM. Please squash the commits.

**Comment by gokulavasan** [comment]
@chtyim Squashed commits. Thanks!

**Comment by hsaputra** [comment]
Thanks for the PR, @gokulavasan. Next time would love to have more information how you solve it in the PR description rather than just link to JIRA.

**Comment by gokulavasan** [comment]
@hsaputra I have added some description to the PR. Sorry. Thank you for catching it.

**Pull request closed without merging**